Test changes from my whinlatter compatibility PR

zboszor opened the pull request:
Hi,

I tried your changes from https://github.com/OSSystems/meta-browser/pull/920 but it failed to build for me, see my comment there.

I cherry-picked/adapted/salvaged some changes from my own PR at https://github.com/OSSystems/meta-browser/pull/918 and it seems to build now.

caneraltinbasak committed:
chromium: Update to 138.0.7204.35

Chromium 138+ uses its bundled libc++ and compiler-rt. To ensure the build
system uses these and does not fall back to system-provided runtime libraries,
explicitly pass the target triple and disable system stdlib++.

This avoids linker errors caused by missing crtbeginS.o, crtendS.o, and libgcc
when using Clang with a minimal sysroot.

Release Notes:
https://chromereleases.googleblog.com/2025/07/stable-channel-update-for-desktop_15.html

Build and patch changes:
Added patches:

0011-fix-check_version-Only-compare-node.js-major-version.patch
Updated patches:

0009-Adjust-the-Rust-build-to-our-needs.patch
Removed patches:

0011-Disable-crabbyavif-to-fix-build-errors.patch
0012-Revert-Allow-and-use-std-hardware_destructive_interf.patch
0013-Revert-Set-Rust-symbol-visibility-to-hidden-when-C-s.patch
0014-pdfium-Fix-missing-PDFiumAPIStringBufferAdapter-temp.patch
0015-Revert-Connect-the-Rust-log-crate-to-the-base-loggin.patch
License changes:
Added licenses:

third_party/compiler-rt/src/LICENSE.TXT
third_party/libtess2/LICENSE
third_party/llvm-libc/src/LICENSE.TXT
third_party/net/third_party/mozilla_security_manager/LICENSE
Removed licenses:

third_party/accessibility-audit/LICENSE
third_party/io_grpc_grpc_api/LICENSE
third_party/libavifinfo/LICENSE
Updated licenses:

third_party/android_deps/libs/com_google_android_gms_play_services_auth/LICENSE → third_party/android_deps/autorolled/committed/libs/com_google_android_gms_play_services_auth/LICENSE
third_party/android_deps/libs/com_google_android_gms_play_services_auth_api_phone/LICENSE → third_party/android_deps/autorolled/committed/libs/com_google_android_gms_play_services_auth_api_phone/LICENSE
third_party/android_deps/libs/com_google_android_gms_play_services_auth_base/LICENSE → third_party/android_deps/autorolled/committed/libs/com_google_android_gms_play_services_auth_base/LICENSE
third_party/android_deps/libs/com_google_android_gms_play_services_auth_blockstore/LICENSE → third_party/android_deps/autorolled/committed/libs/com_google_android_gms_play_services_auth_blockstore/LICENSE
third_party/android_deps/libs/com_google_android_gms_play_services_fido/LICENSE → third_party/android_deps/autorolled/committed/libs/com_google_android_gms_play_services_fido/LICENSE
third_party/android_deps/libs/com_google_android_libraries_identity_googleid_googleid/LICENSE →
third_party/android_deps/autorolled/committed/libs/com_google_android_libraries_identity_googleid_googleid/LICENSE
third_party/expat/src/expat/COPYING
third_party/libjpeg_turbo/LICENSE
Test-built
chromium-ozone-wayland

master, clang, MACHINE=qemux86-64
chromium-x11

master, clang, MACHINE=qemux86-64, qemuarm, qemuarm64
Repo hashes (from meta-chromium-test)
meta-oe:
commit: "a8dfd10600035a799abae03178fc7054582ea43d"
poky:
commit: "243d54fd466f5f852cc0fdcce57997918ce35f32"
meta-clang:
commit: "7a2f83360920b10214e2659e17a4b9cba2d0435b"

zboszor committed:
chromium: 140.0.7339.136

RUNTIME and TC_CXX_RUNTIME are both set explicitly to "llvm"
in chromium.inc. Therefore the :runtime-llvm override is
not needed because it's always on. But it's also a problem in
whinlatter now, because the DEPENDS = "compiler-rt" is not
applied properly, making do_copy_clang_library() fail.
Remove :runtime-llvm override.

Use TOOLCHAIN_NATIVE instead of TOOLCHAIN:class-native.

Drop 0008-Use-the-correct-path-to-libclang_rt.builtins.a.patch
and fix do_copy_clang_library() in a way that's compatible with
clang in oe-core.

Depend on lld-native.

Signed-off-by: Zoltán Böszörményi <zboszor@gmail.com>

caneraltinbasak commented:
I've kicked a build: https://github.com/brightsign/meta-browser/actions/runs/17744437113

caneraltinbasak commented:
Build failed. Can you give me the hashes you used for: meta-oe, meta-clang, poky.

zboszor commented:
Hi,
```
meta-oe    926302f7296299c5adc5b9afa62a6f1ee9424632
meta-clang af59ec636f0f956a94356ede6579e11b8653314d
meta       0a886fcacaab0fbce1306c0f99d482f940a8f705
poky       2ff9be23fffeec94bc5156166ad157b0a59e076d
```

zboszor commented:
I just replaced the last commit. With `meta-clang` merged into `oe-core`, `lld` became a separate recipe, so chromium needs to depend on `lld-native`. My build is running now and does not complain about `-fuse-ld=lld is invalid`.

zboszor commented:
Still failed, with a much advanced state. It complains about a rust module:
```
ld.lld: error: undefined symbol: __rustc::__rust_no_alloc_shim_is_unstable_v2
>>> referenced by serde_json_lenient.8880183df0940bc0-cgu.0
>>>               libserde_json_lenient_lib.serde_json_lenient.8880183df0940bc0-cgu.0.rcgu.o:(serde_json_lenient::error::Error::syntax::h367187de58202ca1) in archive yocto_native/obj/third_party/rust/serde_json_lenient/v0_2/lib/libserde_json_lenient_lib.rlib
>>> referenced by third_uparty_srust_sserde_ujson_ulenient_sv0_u2_swrapper_cwrapper.7bbe238a93e506f6-cgu.0
>>>               libthird_uparty_srust_sserde_ujson_ulenient_sv0_u2_swrapper_cwrapper.third_uparty_srust_sserde_ujson_ulenient_sv0_u2_swrapper_cwrapper.7bbe238a93e506f6-cgu.0.rcgu.o:(_$LT$$RF$mut$u20$serde_json_lenient..de..Deserializer$LT$R$GT$$u20$as$u20$serde..de..Deserializer$GT$::deserialize_any::h57ecaa2d73b56b79) in archive yocto_native/obj/third_party/rust/serde_json_lenient/v0_2/wrapper/libthird_uparty_srust_sserde_ujson_ulenient_sv0_u2_swrapper_cwrapper.rlib
>>> referenced by third_uparty_srust_sserde_ujson_ulenient_sv0_u2_swrapper_cwrapper.7bbe238a93e506f6-cgu.0
>>>               libthird_uparty_srust_sserde_ujson_ulenient_sv0_u2_swrapper_cwrapper.third_uparty_srust_sserde_ujson_ulenient_sv0_u2_swrapper_cwrapper.7bbe238a93e506f6-cgu.0.rcgu.o:(_$LT$$RF$mut$u20$serde_json_lenient..de..Deserializer$LT$R$GT$$u20$as$u20$serde..de..Deserializer$GT$::deserialize_any::h57ecaa2d73b56b79) in archive yocto_native/obj/third_party/rust/serde_json_lenient/v0_2/wrapper/libthird_uparty_srust_sserde_ujson_ulenient_sv0_u2_swrapper_cwrapper.rlib
>>> referenced 144 more times
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

caneraltinbasak commented:
> Hi,
> 
> ```
> meta-oe    926302f7296299c5adc5b9afa62a6f1ee9424632
> meta-clang af59ec636f0f956a94356ede6579e11b8653314d
> meta       0a886fcacaab0fbce1306c0f99d482f940a8f705
> poky       2ff9be23fffeec94bc5156166ad157b0a59e076d
> ```

I've replied your message on other pull request. I'll test this, but I first want to land the existing changes and do the adaptations for latest oe as a follow up. 

Some general comments about the changeset:
1. Conditionals are generally a bad idea in bitbake.
2. You don't need WHINLATTER_OR_NEWER conditional, and it doesn't solve the problem. The meta-oe hash I've picked is also Whinlatter, and doesn't have this problem.
3. meta-browser has branches for older yocto releases. Users shouldn't try to use latest meta-browser with older oe releases anyway.

zboszor commented:
> I've replied your message on other pull request. I'll test this, but I first want to land the existing changes and do the adaptations for latest oe as a follow up.
> 
> Some general comments about the changeset:
> 
>     1. Conditionals are generally a bad idea in bitbake.
> 
>     2. You don't need WHINLATTER_OR_NEWER conditional, and it doesn't solve the problem. The meta-oe hash I've picked is also Whinlatter, and doesn't have this problem.
> 
>     3. meta-browser has branches for older yocto releases. Users shouldn't try to use latest meta-browser with older oe releases anyway.

So the bottom line is meta-browser master should more aggressively remove support for older versions then. This way, no conditionals are necessary.

zboszor commented:
FWIW, I tried to revive my https://github.com/OSSystems/meta-browser/pull/918 with the lld fix, and it also failed with a rust related error, only with more missing symbols that are alloc related. Maybe rust 1.89 needs an even newer chromium, like 139 or 140.

caneraltinbasak commented:
> Still failed, with a much advanced state. It complains about a rust module:
> 
> ```
> ld.lld: error: undefined symbol: __rustc::__rust_no_alloc_shim_is_unstable_v2
Can you please try adding this change as a patch:
https://github.com/chromium/chromium/commit/6aae0e2353c857d98980ff677bf304288d7c58de

zboszor commented:
> Can you please try adding this change as a patch: [chromium/chromium@6aae0e2](https://github.com/chromium/chromium/commit/6aae0e2353c857d98980ff677bf304288d7c58de)

The build is in progress and has long passed the previous failure point. I will report back.

Regarding the extra changes: I can clean up the conditionals and make meta-chromium whinlatter-only.

Or we can keep them, because meta-firefox in the same repository also uses them to keep compatibility with walnascar and older. In fact, the whole conditional idea came from there.

caneraltinbasak commented:
> Regarding the extra changes: I can clean up the conditionals and make meta-chromium whinlatter-only.

This would be my preferred option. Chromium is much more complicated and fragile than Firefox.

zboszor commented:
The build failed at 57% with this:
```
../../sandbox/policy/linux/sandbox_linux.cc:694:3: error: static assertion expression is not an integral constant expression
```

zboszor committed:
meta-chromium: Support whinlatter only

There are breaking changes in whinlatter in oe-core.

Part of the changes was that meta-clang was merged into oe-core.
There's no need to depend on meta-clang anymore.

Signed-off-by: Zoltán Böszörményi <zboszor@gmail.com>

zboszor commented:
> > Regarding the extra changes: I can clean up the conditionals and make meta-chromium whinlatter-only.
> 
> This would be my preferred option. Chromium is much more complicated and fragile than Firefox.

Cleaned up the conditionals.

caneraltinbasak commented:
> The build failed at 57% with this:
> 
> ```
> ../../sandbox/policy/linux/sandbox_linux.cc:694:3: error: static assertion expression is not an integral constant expression
> ```

I suspect this is due to new compiler not being happy with some old code. Can you paste entire error message?

zboszor commented:
```
../../sandbox/policy/linux/sandbox_linux.cc:694:3: error: static assertion expression is not an integral constant expression
  694 |   UMA_HISTOGRAM_ENUMERATION("Security.Sandbox.LandlockState", landlock_state);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../base/metrics/histogram_macros.h:82:7: note: expanded from macro 'UMA_HISTOGRAM_ENUMERATION'
   80 |   INTERNAL_UMA_HISTOGRAM_ENUMERATION_GET_MACRO(                         \
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   81 |       __VA_ARGS__, INTERNAL_UMA_HISTOGRAM_ENUMERATION_SPECIFY_BOUNDARY, \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   82 |       INTERNAL_UMA_HISTOGRAM_ENUMERATION_DEDUCE_BOUNDARY)               \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   83 |   (name, __VA_ARGS__, base::HistogramBase::kUmaTargetedHistogramFlag)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../base/metrics/histogram_macros_internal.h:171:73: note: expanded from macro 'INTERNAL_UMA_HISTOGRAM_ENUMERATION_GET_MACRO'
  171 | #define INTERNAL_UMA_HISTOGRAM_ENUMERATION_GET_MACRO(_1, _2, NAME, ...) NAME
      |                                                                         ^
../../base/metrics/histogram_macros_internal.h:175:3: note: expanded from macro 'INTERNAL_UMA_HISTOGRAM_ENUMERATION_DEDUCE_BOUNDARY'
  175 |   INTERNAL_HISTOGRAM_ENUMERATION_WITH_FLAG(                                    \
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  176 |       name, sample,                                                            \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  177 |       base::internal::EnumSizeTraits<std::decay_t<decltype(sample)>>::Count(), \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  178 |       flags)
      |       ~~~~~~
../../base/metrics/histogram_macros_internal.h:207:9: note: expanded from macro 'INTERNAL_HISTOGRAM_ENUMERATION_WITH_FLAG'
  207 |         static_cast<uintmax_t>(boundary) <                                     \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  208 |             static_cast<uintmax_t>(                                            \
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  209 |                 std::numeric_limits<base::HistogramBase::Sample32>::max()),    \
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../base/metrics/histogram_macros_internal.h:35:14: note: integer value 4 is outside the valid range of values [0, 3] for the enumeration type 'LandlockState'
   35 |       return static_cast<Enum>(base::to_underlying(Enum::kMaxValue) + 1);
      |              ^
../../sandbox/policy/linux/sandbox_linux.cc:694:3: note: in call to 'Count()'
  694 |   UMA_HISTOGRAM_ENUMERATION("Security.Sandbox.LandlockState", landlock_state);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../base/metrics/histogram_macros.h:82:7: note: expanded from macro 'UMA_HISTOGRAM_ENUMERATION'
   80 |   INTERNAL_UMA_HISTOGRAM_ENUMERATION_GET_MACRO(                         \
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   81 |       __VA_ARGS__, INTERNAL_UMA_HISTOGRAM_ENUMERATION_SPECIFY_BOUNDARY, \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   82 |       INTERNAL_UMA_HISTOGRAM_ENUMERATION_DEDUCE_BOUNDARY)               \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   83 |   (name, __VA_ARGS__, base::HistogramBase::kUmaTargetedHistogramFlag)
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../base/metrics/histogram_macros_internal.h:171:73: note: expanded from macro 'INTERNAL_UMA_HISTOGRAM_ENUMERATION_GET_MACRO'
  171 | #define INTERNAL_UMA_HISTOGRAM_ENUMERATION_GET_MACRO(_1, _2, NAME, ...) NAME
      |                                                                         ^
../../base/metrics/histogram_macros_internal.h:177:7: note: expanded from macro 'INTERNAL_UMA_HISTOGRAM_ENUMERATION_DEDUCE_BOUNDARY'
  175 |   INTERNAL_HISTOGRAM_ENUMERATION_WITH_FLAG(                                    \
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  176 |       name, sample,                                                            \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  177 |       base::internal::EnumSizeTraits<std::decay_t<decltype(sample)>>::Count(), \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  178 |       flags)
      |       ~~~~~~
../../base/metrics/histogram_macros_internal.h:207:32: note: expanded from macro 'INTERNAL_HISTOGRAM_ENUMERATION_WITH_FLAG'
  207 |         static_cast<uintmax_t>(boundary) <                                     \
      |                                ^~~~~~~~
1 error generated.
```

caneraltinbasak commented:
it is due to strict argument number checking for macros on Clang21. You need to add this patch to your build:
https://github.com/chromium/chromium/commit/b0ff8c3b258a8816c05bdebf472dbba719d3c491

zboszor commented:
Compiling `sandbox/policy/linux/sandbox_linux.cc` succeeds with that patch. Still waiting for the build to finish. I will report back.

zboszor commented:
Next error, this time it's Rust again:
```
FAILED: [code=1] obj/third_party/rust/qr_code/v2/lib/libqr_code_lib.rlib 
"python3" "../../build/rust/rustc_wrapper.py" --rustc=../../../../recipe-sysroot-native/usr/bin/rustc --depfile=obj/third_party/rust/qr_code/v2/lib/libqr_code_lib.rlib.d 
 --rsp=obj/third_party/rust/qr_code/v2/lib/libqr_code_lib.rlib.rsp -- -Clinker="x86_64-oe-linux-clang++  -m64 -march=nehalem -mtune=generic -mfpmath=sse -msse4.2  --dyld
-prefix=/usr -fstack-protector-strong  -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/mnt/zozo/yocto-5.3-arm/tmp-sicom/work/corei7
-64-oe-linux/chromium-x11/138.0.7204.157/recipe-sysroot" --crate-name qr_code ../../third_party/rust/chromium_crates_io/vendor/qr_code-v2/src/lib.rs --crate-type rlib -A
warnings -Cmetadata=qr_code-2 --edition=2018 -Cforce-unwind-tables=no -Crelocation-model=pic -Cembed-bitcode=no -Cdefault-linker-libraries -Zdep-info-omit-d-target -Zmac
ro-backtrace -Zremap-cwd-prefix=. -Zexternal-clangrt --color=always --target=x86_64-oe-linux-gnu -Cembed-bitcode=no -Clto=no -Ccodegen-units=1 --cfg cr_rustc_revision=\"
custom\" -Copt-level=3 -Zallow-features= -Cembed-bitcode=no -Zdefault-visibility=hidden -Funsafe_code --emit=dep-info=obj/third_party/rust/qr_code/v2/lib/libqr_code_lib.
rlib.d,link -o obj/third_party/rust/qr_code/v2/lib/libqr_code_lib.rlib LDFLAGS RUSTENV OUT_DIR=../../../../../../out/Release/gen/third_party/rust/qr_code/v2/lib CARGO_PK
G_AUTHORS=kennytm\ \<kennytm@gmail.com\>,\ Riccardo\ Casatta\ \<riccardo.casatta@gmail.com\> CARGO_PKG_VERSION=2.0.0 CARGO_PKG_NAME=qr_code CARGO_PKG_DESCRIPTION=QR\ cod
e\ encoder\ in\ Rust,\ support\ structured\ append\ \(data\ in\ multiple\ qrcodes\) CARGO_MANIFEST_DIR=../../third_party/rust/qr_code/v2/lib/crate

error: hiding a lifetime that's elided elsewhere is confusing
   --> ../../third_party/rust/chromium_crates_io/vendor/qr_code-v2/src/lib.rs:219:17
    |
219 |     pub fn iter(&self) -> QrCodeIterator {
    |                 ^^^^^     -------------- the same lifetime is hidden here
    |                 |
    |                 the lifetime is elided here
    |
    = help: the same lifetime is referred to in inconsistent ways, making the signature confusing
note: the lint level is defined here
   --> ../../third_party/rust/chromium_crates_io/vendor/qr_code-v2/src/lib.rs:8:9
    |
8   | #![deny(warnings)]
    |         ^^^^^^^^
    = note: `#[deny(mismatched_lifetime_syntaxes)]` implied by `#[deny(warnings)]`
help: use `'_` for type paths
    |
219 |     pub fn iter(&self) -> QrCodeIterator<'_> {
    |                                         ++++

error: hiding a lifetime that's elided elsewhere is confusing
   --> ../../third_party/rust/chromium_crates_io/vendor/qr_code-v2/src/optimize.rs:112:22
    |
112 |     pub fn new(data: &[u8]) -> Parser {
    |                      ^^^^^     ------ the same lifetime is hidden here
    |                      |
    |                      the lifetime is elided here
    |
    = help: the same lifetime is referred to in inconsistent ways, making the signature confusing
help: use `'_` for type paths
    |
112 |     pub fn new(data: &[u8]) -> Parser<'_> {
    |                                      ++++

error: aborting due to 2 previous errors
```

caneraltinbasak commented:
Seems like Chromium haven't started using the latest Rust yet. I can't find a patch addressing this issue upstream on latest Chromium. You need these changes in the source:

```
diff --git a/third_party/rust/chromium_crates_io/vendor/qr_code-v2/src/lib.rs b/third_party/rust/chromium_crates_io/vendor/qr_code-v2/src/lib.rs
@@ -216,7 +216,7 @@ impl QrCode {
-    pub fn iter(&self) -> QrCodeIterator {
+    pub fn iter(&self) -> QrCodeIterator<'_> {
         QrCodeIterator::new(self)
     }
 }
```
and
 
```
diff --git a/third_party/rust/chromium_crates_io/vendor/qr_code-v2/src/optimize.rs b/third_party/rust/chromium_crates_io/vendor/qr_code-v2/src/optimize.rs
@@ -109,7 +109,7 @@ impl Parser {
-    pub fn new(data: &[u8]) -> Parser {
+    pub fn new(data: &[u8]) -> Parser<'_> {
         // existing body
     }
 }
```

You will likely find more problems though.

zboszor commented:
I went a different route. Instead of finding patches that can be cherry picked, I upgraded to chromium 140 and it completed the build.

zboszor committed:
chromium: Add INSANE_SKIP = "buildpaths" to the debug package

Signed-off-by: Zoltán Böszörményi <zboszor@gmail.com>
